forge/src/main/java/me/lucko/luckperms/forge/service/ForgePermissionHandler.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * This file is part of LuckPerms, licensed under the MIT License.
+ *
+ *  Copyright (c) lucko (Luck) <luck@lucko.me>
+ *  Copyright (c) contributors
+ *
+ *  Permission is hereby granted, free of charge, to any person obtaining a copy
+ *  of this software and associated documentation files (the "Software"), to deal
+ *  in the Software without restriction, including without limitation the rights
+ *  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ *  copies of the Software, and to permit persons to whom the Software is
+ *  furnished to do so, subject to the following conditions:
+ *
+ *  The above copyright notice and this permission notice shall be included in all
+ *  copies or substantial portions of the Software.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ *  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ *  SOFTWARE.
+ */
+
+package me.lucko.luckperms.forge.service;
+
+import me.lucko.luckperms.forge.LPForgeBootstrap;
+import me.lucko.luckperms.forge.bridge.server.level.ServerPlayerBridge;
+import net.luckperms.api.util.Tristate;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.server.level.ServerPlayer;
+import net.minecraftforge.server.permission.handler.IPermissionHandler;
+import net.minecraftforge.server.permission.nodes.PermissionDynamicContext;
+import net.minecraftforge.server.permission.nodes.PermissionNode;
+import net.minecraftforge.server.permission.nodes.PermissionTypes;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.UUID;
+
+public class ForgePermissionHandler implements IPermissionHandler {
```

Comment:
Can we support contexts here?

Additionally, we should be able to support "meta" (prefix, suffix etc) lookups for the other permissiontypes.

Finally, can probably support offline permission lookup if the user data is cached.

Reply:
I've added context and offline permission / meta support, not entirely sure if this is the correct way to add contexts to a existing `QueryOptions`
https://github.com/LuckPerms/LuckPerms/blob/fed94e36fd55a919dff75a0a4d0b3e063f1d480d/forge/src/main/java/me/lucko/luckperms/forge/service/ForgePermissionHandler.java#L125-L134

Verbose:
- Online: https://luckperms.net/verbose/wyDwvVie8t
- Offline: https://luckperms.net/verbose/l80gEzDrxg